handler/OpenIDHandler.php

```diff
@@ -0,0 +1,464 @@
+            $templateMgr->display('frontend/pages/userRegisterComplete.tpl');
+        } elseif (!$request->isPost()) {
+            $request->redirect($context->getPath(), 'login');
+        } else {
+            $plugin = PluginRegistry::getPlugin('generic', KEYCLOAK_PLUGIN_NAME);
+            $regForm = new OpenIDStep2Form($plugin);
+            $regForm->readInputData();
+            if (!$regForm->validate()) {
+                $regForm->display($request);
+            } elseif ($regForm->execute()) {
+                $request->redirect($context->getPath(), 'openid', 'registerOrConnect');
+            } else {
+                $regForm->addError('', '');
+                $regForm->display($request);
+            }
+        }
+    }
+
+    public static function updateUserDetails($payload, $user, $request, $selectedProvider, $setProviderId = false)
+    {
+        $context = $request->getContext();
+        $contextId = ($context == null) ? 0 : $context->getId();
+        $plugin = PluginRegistry::getPlugin('generic', KEYCLOAK_PLUGIN_NAME);
+        $settings = json_decode($plugin->getSetting($contextId, 'openIDSettings'), true);
+
+        if (key_exists('providerSync', $settings) && $settings['providerSync'] == 1) {
+            $site = $request->getSite();
+            $sitePrimaryLocale = $site->getPrimaryLocale();
+            $currentLocale = Locale::getLocale();
+            if (is_array($payload) && key_exists('given_name', $payload) && !empty($payload['given_name'])) {
+                $user->setGivenName($payload['given_name'], ($sitePrimaryLocale != $currentLocale) ? $sitePrimaryLocale : $currentLocale);
+            }
+            if (is_array($payload) && key_exists('family_name', $payload) && !empty($payload['family_name'])) {
+                $user->setFamilyName($payload['family_name'], ($sitePrimaryLocale != $currentLocale) ? $sitePrimaryLocale : $currentLocale);
+            }
+            if (is_array($payload) && key_exists('email', $payload) && !empty($payload['email']) && Repo::user()->getByEmail($payload['email']) == null) {
+                $user->setEmail($payload['email']);
+            }
+            if ($selectedProvider == 'orcid') {
+                if (is_array($payload) && key_exists('id', $payload) && !empty($payload['id'])) {
+                    $user->setOrcid($payload['id']);
+                }
+            }
+        }
+
+        $user->setData('openid::lastProvider', $selectedProvider);
+
+        if (is_array($payload) && key_exists('id', $payload) && !empty($payload['id'])) {
+            if ($setProviderId) {
+                $user->setData('openid::'.$selectedProvider, $payload['id']);
+            }
+            $generateApiKey = isset($settings) && key_exists('generateAPIKey', $settings) ? $settings['generateAPIKey'] : false;
+            $secret = Config::getVar('security', 'api_key_secret', '');
+            if ($generateApiKey && $selectedProvider == 'custom' && $secret) {
```

Comment:
It might be a good idea to pass something into the erorr log if the API key secret is empty.

Reply:
Changed it to 

```
$secret = Config::getVar('security', 'api_key_secret');

            if (!isset($secret)) {
                error_log($plugin->getName() . ' - api_key_secret not defined in configuration file');
            } else if ($generateApiKey && $selectedProvider == OpenIDPlugin::CUSTOM_PROVIDER) {
                $user->setData('apiKeyEnabled', true);
                $user->setData('apiKey', self::encryptOrDecrypt($plugin, $contextId, 'encrypt', $payload['id']));
                Repo::user()->edit($user);
            }
```